ML-DSA FIPS 204

strongX509 opened the pull request:
The ml-dsa branch implements the ML-DSA post-quantum signature algorithms MS-DSA-44, MS-DSA-65 and MS-DS-87 as specified by the FIPS 204 standard. Either the botan or wolfssl crypto libraries can be used by attaching the botan and wolfssl plugins, respectively. In addition to that we offer our own reference implementation  via the ml plugin.

strongX509 committed:
libstrongswan: Add definitions for ML-DSA keys and signatures

strongX509 committed:
wolfssl: Add ML-DSA support

strongX509 committed:
pki: Add ML-DSA support

strongX509 committed:
scripts: Added nist_acvp script

strongX509 committed:
unit-tests: Added ml_dsa test suite

strongX509 committed:
botan: Add ML-DSA support

strongX509 committed:
curve25519: Use of public_key_info_encode()

strongX509 committed:
ml: Add ML-DSA support

strongX509 committed:
ml: Bug fix in ml_bitpacker

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_ml_dsa_private_key.c:
```diff
@@ -0,0 +1,296 @@
+{
+	char *sig_mode;
+
+	sig_mode = 	(lib->settings->get_bool(lib->settings, "%s.ml_dsa.deterministic",
+						FALSE, lib->ns)) ? "Deterministic" : "Randomized";
```
```suggestion
	sig_mode = lib->settings->get_bool(lib->settings, "%s.ml_dsa.deterministic",
							FALSE, lib->ns) ? "Deterministic" : "Randomized";
```

strongX509 replied:
Fixed

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_ml_dsa_private_key.c:
```diff
@@ -0,0 +1,296 @@
+	this = create_empty(type);
+
+	if (botan_privkey_create(&this->key, "ML-DSA",
+			botan_ml_dsa_get_mldsa_mode(type), rng))
```
```suggestion
							 botan_ml_dsa_get_mldsa_mode(type), rng))
```

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_ml_dsa_private_key.c:
```diff
@@ -0,0 +1,296 @@
+	this = create_empty(type);
+
+	if (botan_privkey_load_ml_dsa(&this->key, priv.ptr, priv.len,
+			botan_ml_dsa_get_mldsa_mode(type)))
```
```suggestion
								  botan_ml_dsa_get_mldsa_mode(type)))
```

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,225 @@
+		.type = type,
+		.ref = 1,
+	);
+
```
```suggestion
```

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,225 @@
+ */
+ const char *botan_ml_dsa_get_mldsa_mode(key_type_t type)
+ {
+
```
```suggestion
```

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,225 @@
+		}
+		break;
+	}
+
```
```suggestion
```

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,225 @@
+	this = create_empty(type);
+
+	if (botan_pubkey_load_ml_dsa(&this->key, key.ptr, key.len,
+			botan_ml_dsa_get_mldsa_mode(type)))
```
```suggestion
								 botan_ml_dsa_get_mldsa_mode(type)))
```

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_util_keys.c:
```diff
@@ -128,6 +130,18 @@ public_key_t *botan_public_key_load(key_type_t type, va_list args)
+	{
+		if (type == KEY_ANY)
+		{
+			type = public_key_info_decode(blob, NULL);
```
What happens if the type can't be determined? `botan_ml_dsa_public_key_adopt()` should probably fail if `type` is `KEY_ANY`.

strongX509 replied:
botan parses the public key already in line 92 and exits if the key type is not supported:
```
	if (botan_pubkey_load(&pubkey, blob.ptr, blob.len))
	{
		botan_rng_destroy(rng);
		return NULL;
	}
```
Our problem is that the botan **`pubkey`** object currently doesn't offer a method to extract which type of ML-DSA key was loaded (according to the source code this is on the TODO list for botan 4). Therefore we have to do our own PKCS#1 parsing to extract the key type (one of the reasons for the **`public_key_info_decode()`**  function)

tobiasbrunner replied:
OK, I guess it's unlikely that `"ML-DSA"` keys are of a (sub-)type we don't support (i.e. don't know the OID of - not sure if that could be an issue in the future, though). Would it make sense to check if the type is correct? That is, always decode the blob and then compare the type if it's not `KEY_ANY`? Or just always use the decoded type if it's known?

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_util_keys.c:
```diff
@@ -232,6 +265,18 @@ private_key_t *botan_private_key_load(key_type_t type, va_list args)
+#endif
+#ifdef BOTAN_HAS_ML_DSA
+	if (streq(name, "ML-DSA") && (type == KEY_ANY || type == KEY_ML_DSA_44 ||
+							  type == KEY_ML_DSA_65 || type == KEY_ML_DSA_87))
```
```suggestion
								  type == KEY_ML_DSA_65 || type == KEY_ML_DSA_87))
```

tobiasbrunner replied:
Actually, this one is still the same, the public key one above was possibly fixed.

tobiasbrunner commented on src/libstrongswan/plugins/botan/botan_util_keys.c:
```diff
@@ -232,6 +265,18 @@ private_key_t *botan_private_key_load(key_type_t type, va_list args)
+	{
+		if (type == KEY_ANY)
+		{
+			type = key_type_from_pkcs8(blob);
```
What if the type can't be determined? `botan_ml_dsa_private_key_adopt()` should probably fail if `type` is `KEY_ANY`.

strongX509 replied:
botan parses the public key already in line 255 and exits if the key type is not supported:
```
	if (botan_privkey_load(&key, rng, blob.ptr, blob.len, NULL))
	{
		botan_rng_destroy(rng);
		return NULL;
	}
```
Our problem is that the botan **`key`** object currently doesn't offer a method to extract which type of ML-DSA key was loaded (according to the source code this is on the TODO list for botan 4). Therefore we have to do our own PKCS#8 parsing to determine the key type. Since the **`ml`** and **`wolfssl`** plugins are currently not able to do their own PKCS#8 parsing of an ASN.1 encoded ML-DSA private key, a **`private_key_info_decode()`** function would be practical with the long term goal of getting rid of the **`pkcs1`** and **`pkcs8`** plugins, altoghether.

tobiasbrunner replied:
Similar to the discussion above with the public key, would it make sense to compare or adopt the actually parsed type (if it's not `KEY_ANY`, just to make this future-proof if there will ever be other OIDs)?

tobiasbrunner commented on src/libstrongswan/credentials/keys/public_key.c:
```diff
@@ -109,6 +117,116 @@ bool public_key_has_fingerprint(public_key_t *public, chunk_t fingerprint)
+		{
+			case PKINFO_SUBJECT_PUBLIC_KEY_ALGORITHM:
+				oid = asn1_parse_algorithmIdentifier(object,
+						parser->get_level(parser)+1, NULL);
```
```suggestion
											parser->get_level(parser)+1, NULL);
```

strongX509 replied:
Changed indentation

tobiasbrunner reviewed (changes requested):
Thanks a lot. So far I've reviewed the changes up to and including those to the `botan` plugin (c8aa75834f28f14c15ed3874c77e55acbc6e02ee). I'll have a look at the `ml` plugin next week. Or would it maybe be better to create a separate PR for it (reviewing it might take a while)?

strongX509 commented on src/libstrongswan/credentials/keys/public_key.h:
```diff
@@ -285,5 +322,12 @@ enumerator_t *signature_schemes_for_key(key_type_t type, int size);
 key_type_t key_type_from_signature_scheme(signature_scheme_t scheme);
 
+/**
+ * Return the public key size in octets for post-quantum algorithms
```
Initially it was intended for post-quantum keys only. But the function could also be used in the EdDSA case, replacing some internally defined constants. This hasn't been done yet. For RSA and ECDSA keys the function is not of much use since the key sizes vary.

Currently **`get_public_key_size()`** is used in the following places:
~/strongswan/src/libstrongswan/plugins/botan/botan_ml_dsa_private_key.c:
~/strongswan/src/libstrongswan/plugins/botan/botan_ml_dsa_public_key.c:
~/strongswan/src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
~/strongswan/src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
~/strongswan/src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_private_key.c:
~/strongswan/src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_public_key.c:

strongX509 commented on src/libstrongswan/plugins/pkcs1/pkcs1_builder.c:
```diff
@@ -100,12 +112,20 @@ static public_key_t *parse_public_key(chunk_t blob)
+							break;
+						default:
+							key = lib->creds->create(lib->creds, CRED_PUBLIC_KEY,
+									 type, BUILD_BLOB, object, BUILD_END);
```
The final version of **`parse_public_key()`** looks radically different and got rid of the DBG2 statement altogether.  The intention is to treat EdDSA keys the same way as ML-DSA keys in a next step and forwarding them as blobs with the PKCS#1 header stripped from them, so that they don't have to be parsed a second time. The **`switch`** statement will still be necessary since ECDSA key types still have to be forwarded with the PKCS#1 header intact.

tobiasbrunner replied:
Yes, I saw that refactoring. However, I didn't see my comment anymore during my second review session. I wonder if it got hidden due to the push to the branch (or maybe it was moved somewhere I didn't look).


> The intention is to treat EdDSA keys the same way as ML-DSA keys in a next step and forwarding them as blobs with the PKCS#1 header stripped from them, so that they don't have to be parsed a second time.

They are not really the same, though. The problem with them is that they are double-wrapped in OCTET STRINGs. Or are you intending to do that unwrapping as well?

strongX509 replied:
No, the second octet string is not part of the PKCS#1 encoding. Thus the unwrapping should be done by the plugins.

strongX509 commented on src/pki/commands/gen.c:
```diff
@@ -112,6 +124,15 @@ static int gen()
 			case KEY_ED448:
 				size = 456;
 				break;
+			case KEY_ML_DSA_44:
+				size = 10496;
+				break;
+			case KEY_ML_DSA_65:
+				size = 15616;
+				break;
+			case KEY_ML_DSA_87:
+				size = 20736;
+				break;
```
The output public key size is just informative in order to frighten the users when they compare it to RSA key sizes but doesn't have any practical use. Better than key size 0 which would have to be suppressed. An alternative would be effective cryptographic strength but nobody is quite sure about the actual strength of ML-based algorithms.

tobiasbrunner replied:
> The output public key size is just informative in order to frighten the users when they compare it to RSA key sizes but doesn't have any practical use.

It's not used, though. It's passed to the builders, which will ignore it for EdDSA and ML-DSA. The `--gen` command doesn't use the `size` variable for anything else.

strongX509 replied:
Deleted

strongX509 commented on conf/options/charon.opt:
```diff
@@ -301,6 +301,14 @@ charon.make_before_break = yes
 	gaps during reauthentication, but requires support for overlapping SAs by
 	the peer. strongSwan can handle such overlapping SAs since version 5.3.0.
 
+charon.ml_dsa.context
+	Optional context string with a maximum length of 255 bytes. Either a text
+	string or hex-encoded binary bytes with a 0x prefix.
```
The use of context strings is mentionend in FIPS-204 as the big difference between the internal and the external ML-DSA sign function. A context header is always prepended to the message to be signed, even if the context string has zero length (ctx prefix 0x00 0x00). I would have liked to include a signing example with a non-zero context string but unfortunately botan has this feature on its TODO list and offers a hard-coded empty context string, only.

During initial development I worked with test data including a context string and was able to verify the correctness of the  wolfssl and ml implementions.

tobiasbrunner replied:
As discussed, we could implement it where it's supported and add test cases that use `warn()`, similar to the RSA-PSS tests with custom salt (there is `pss_salt_test()`, which tests for the feature and then a warning is printed if it's not supported).

strongX509 replied:
If context strings are not supported (currently botan) a warning is issued.

tobiasbrunner replied:
OK, thanks. We might have to reconsider the context strings anyway if they are e.g. different for IKEv2 and the PKI etc. (might have to be passed as parameters to the signature generation/verification).

strongX509 commented on conf/options/charon.opt:
```diff
@@ -301,6 +301,14 @@ charon.make_before_break = yes
+
+charon.ml_dsa.deterministic = no
+	Enables deterministic ML-DSA signatures primarily used for test purposes.
+	The	recommended default settinga are randomized ML-DSA signatures.
```
Fixed (bad eye sight due to old age again interfering)

tobiasbrunner replied:
:eyeglasses: ;)

strongX509 commented on scripts/nist_acvp.c:
```diff
@@ -0,0 +1,183 @@
+	fprintf(out, "%s [OPTIONS]\n\n", name);
+	fprintf(out, "Options:\n");
+	fprintf(out, "  -h, --help          print this help.\n");
+	fprintf(out, "  -i, --in=FILE       request file (default STDIN).\n");
+	fprintf(out, "  -o, --out=FILE      response file (default STDOUT).\n");
```
changed to input/output file

strongX509 commented on scripts/nist_acvp.c:
```diff
@@ -0,0 +1,183 @@
+		param = line;
+		param_len = pos - line;
+
+		/*remove preceding whitespace from param */
```
Fixed

strongX509 commented on scripts/nist_acvp.c:
```diff
@@ -0,0 +1,183 @@
+		}
+
+		/* remove double quotes from param */
+		if (*param != '"' || param[param_len - 1] != '"')
```
Fixed

strongX509 commented on scripts/nist_acvp.c:
```diff
@@ -0,0 +1,183 @@
+		/* remove double quotes from param */
+		if (*param != '"' || param[param_len - 1] != '"')
+		{
+			fprintf(stderr, "no double quotes around found\n");
```
Fixed

strongX509 commented on src/libstrongswan/tests/Makefile.am:
```diff
@@ -66,7 +66,8 @@ libstrongswan_tests_SOURCES = tests.h tests.c \
   suites/test_ed25519.c \
   suites/test_ed448.c \
-  suites/test_signature_params.c \
+   suites/test_ml_dsa.c \
+ suites/test_signature_params.c \
```
Fixed

strongX509 commented on src/libstrongswan/tests/suites/test_ml_dsa.c:
```diff
@@ -0,0 +1,1353 @@
+	ck_assert(key == NULL);
+
+	/* load private key */
+	key = lib->creds->create(lib->creds, CRED_PRIVATE_KEY, tests[_i].type,
+							 BUILD_PRIV_ASN1_DER, tests[_i].seed, BUILD_END);
```
Actually the naming of the new type `BUILD_PRIV_ASN1_DER` type that I somehow took from the old  **`six-beta`** branch is misleading since the private key stripped from its PKCS#8 wrapper doesn't contain any ASN.1 encoding (This is the new brave world of transferring integers in little endian format in modern crypto algorithms). I change to existing type `BUILD_BLOB`  to be in alignment with the public key decoding use case.

Additionally I added `BUILD_BLOB_ASN1_DER` private and public key load test cases.

strongX509 replied:
Both ASN.1 wrapped and plain private key blobs are now tested. `BUILD_BLOB` is now used.

tobiasbrunner replied:
Looks good, thanks!

strongX509 commented on src/libstrongswan/tests/suites/test_ml_dsa.c:
```diff
@@ -0,0 +1,1353 @@
+
+	/* load public key */
+	pubkey = lib->creds->create(lib->creds, CRED_PUBLIC_KEY, tests[_i].type,
+								BUILD_BLOB, tests[_i].pubkey, BUILD_END);
```
Added test case - see above

strongX509 replied:
see above

strongX509 commented on src/libstrongswan/tests/suites/test_ml_dsa.c:
```diff
@@ -0,0 +1,1353 @@
+							lib->ns);
+
+	/* set empty context */
+	lib->settings->set_str(lib->settings, "%s.ml_dsa.context", "", lib->ns);
```
Yes. I didn't know if `NULL`  was a valid argument with **`set_str()`**. This statement just makes sure that any existing context statement in **`strongswan.conf`** will be overridden as was the case when I was building the **`ml`** plugin code. I don't think that any IKEv2 RFC is going to postulate a context string, but some corporate users might wish to use one (**`context="Elon Musk prefers X`**").

tobiasbrunner replied:
> I didn't know if `NULL` was a valid argument with **`set_str()`**.

That's not exactly what I meant. When you use `get_str()`, you pass a default value. If nothing is configured (i.e. the searched key/value pair does not exist), that's what you get back. However, setting `""` here, will store that value and you get that empty string back. So it depends on what defaults the plugins use (if the default is `""`, then there is no difference, if it's NULL it depends on how that's treated).

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_private_key.c:
```diff
@@ -0,0 +1,385 @@
+	sig_len = wc_dilithium_sig_size(&this->key);
+	*signature = chunk_alloc(sig_len);
+
+	/* deterministic or randomized signature? */
+	if 	(lib->settings->get_bool(lib->settings, "%s.ml_dsa.deterministic", FALSE,
+								 lib->ns))
```
The **`ml`** plugin has its internal **`ml_dsa_get_context()`** function which is quite involved since both plaintext and binary context strings must be supported. A similar function would be needed for the **`wolfssl`** plugin. For the time being I resigned from implementing it, waiting for context string support in the **`botan`** libray to be implemented.

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_private_key.c:
```diff
@@ -0,0 +1,385 @@
+
+	if (ret != 0)
+	{
+		DBG1(DBG_LIB, "signature with scheme %N failed",
+					   signature_scheme_names, scheme);
```
Hmm, this is the default error response of all currently implemented signature schemes.

tobiasbrunner replied:
Hm, I have not found any other such messages in the `wolfssl` plugin.

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,355 @@
+	bool success = FALSE;
+
+	*fp = chunk_empty;
+	len = get_public_key_size(type);
```
The **`wolfssl`** dilithium source code has been partially but not very consistently (using camel case notation) casted to ML-DSA functions and constants but the names might change anytime. The **`botan`** ffi.h C API doesn't support retrieving the public key size of the ML-DSA variants at all and the **`ml`** plugin would have to define the public key sizes in its internal **`params`** struct. This is quite a PITA. I'd prefer a global function which could be used to assign memory or check the size of a public key blob in advance, independent of the ML-DSA implementation that is going to be used.

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,355 @@
+METHOD(public_key_t, get_keysize, int,
+	private_public_key_t *this)
+{
+	return BITS_PER_BYTE * get_public_key_size(this->type);
```
See comment below. I'd prefer a global function based on FIPS 204-defined key sizes and not multiple locally defined constants.

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,355 @@
+			break;
+		default:
+			goto end;
+		}
```
Fixed

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,355 @@
+	chunk_t pubkey;
+	int len, oid;
+
+	len = get_public_key_size(this->type);
```
See comment above.

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_public_key.c:
```diff
@@ -0,0 +1,355 @@
+			DBG1(DBG_LIB, "decoding ML-DSA public key failed");
+			destroy(this);
+			return NULL;
+		}
```
This code has been removed. PKCS#1 decoding is know done by the global **`public_key_info_decode`** function. Why use library code if we have our own.

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_plugin.c:
```diff
@@ -515,6 +517,62 @@ METHOD(plugin_t, get_features, int,
+	#ifndef WOLFSSL_NO_ML_DSA_87
+			PLUGIN_PROVIDE(PUBKEY, KEY_ML_DSA_87),
+	#endif
+			PLUGIN_PROVIDE(PUBKEY, KEY_ANY),
```
See comment above.

strongX509 commented on src/libstrongswan/plugins/wolfssl/wolfssl_ml_dsa_private_key.c:
```diff
@@ -0,0 +1,385 @@
+	}
+
+	public_key = lib->creds->create(lib->creds, CRED_PUBLIC_KEY, this->type,
+								 BUILD_BLOB, pubkey, BUILD_END);
```
Fixed

strongX509 commented on src/libstrongswan/credentials/keys/public_key.c:
```diff
@@ -109,6 +117,116 @@ bool public_key_has_fingerprint(public_key_t *public, chunk_t fingerprint)
+
+	return success ? key_type_from_oid(oid) : KEY_ANY;
+}
+
+
```
Removed

strongX509 commented on src/libstrongswan/credentials/keys/public_key.h:
```diff
@@ -251,6 +263,40 @@ bool public_key_equals(public_key_t *public, public_key_t *other);
+/**
+ * ASN.1 encoding of public key info
+ *
+ * @param pkcs1			ASB,1 encoded public in PCKS#1 format
```
Fixed

tobiasbrunner commented on src/libstrongswan/plugins/ml/ml_plugin.c:
```diff
@@ -62,6 +64,25 @@ METHOD(plugin_t, get_features, int,
+		PLUGIN_PROVIDE(PRIVKEY_SIGN, SIGN_ML_DSA_87),
+		PLUGIN_PROVIDE(PUBKEY_VERIFY, SIGN_ML_DSA_44),
+		PLUGIN_PROVIDE(PUBKEY_VERIFY, SIGN_ML_DSA_65),
+		PLUGIN_PROVIDE(PUBKEY_VERIFY, SIGN_ML_DSA_87),
```
These `PLUGIN_PROVIDE()` features should technically all declare the plugin features on which they depend. Because if e.g. one of the XOFs or RNG classes is missing, the features here would still be registered by the plugin but then wouldn't work at runtime. For credential features (as compared to algorithm features) and if everything is checked in the constructor that's kinda OK, as there could still be a fallback on another implementation to parse/load/generate the key. But it's bad if a dependency is only necessary later and causing some operation to fail while another implementation would have worked correctly (e.g. the SHA-1 dependency for fingerprinting or the `RNG_STRONG` dependency to generate signatures).

Dependencies can also have an impact on the load order. Just as an example, if a plugin tried to load an ML-DSA key in its initialization code (correctly declaring that dependency), this would only work if all the required dependencies are already loaded. Only if these are provided by a plugin that's loaded before the `ml` plugin would that be the case if the dependencies are not declared (if they are, those features are loaded recursively first until all dependencies are resolved or the feature not loaded due to missing dependencies). Here this mostly wouldn't be an issue because the declared ML-KEM features, which are loaded first, have basically the same dependencies, so those would already be loaded if they are available (the two exceptions are SHA-1 for the public key and `RNG_TRUE` for the `PRIVKEY_GEN` features). However, if SHA-3 is not available, the XOFs would not get resolved because the ML-KEM features have already failed loading.

So it would definitely be better to declare the features properly. Maybe some macros could define the dependencies for public and private key features to avoid some repetition?

```c
#define ML_DSA_DEPS \
	PLUGIN_DEPENDS(XOF, XOF_SHAKE_128), \
	PLUGIN_DEPENDS(XOF, XOF_SHAKE_256)

#define ML_DSA_PUBKEY_DEPS \
	ML_DSA_DEPS, \
	PLUGIN_DEPENDS(HASHER, HASH_SHA1)

#define ML_DSA_PRIVKEY_DEPS \
	ML_DSA_DEPS, \
	PLUGIN_DEPENDS(RNG, RNG_STRONG)

#define ML_DSA_PRIVKEY_GEN_DEPS \
	ML_DSA_PRIVKEY_DEPS, \
	PLUGIN_DEPENDS(RNG, RNG_TRUE)
```
Then this would then look like this:
```c
		PLUGIN_REGISTER(PUBKEY, ml_dsa_public_key_load, TRUE),
			PLUGIN_PROVIDE(PUBKEY, KEY_ML_DSA_44),
				ML_DSA_PUBKEY_DEPS,
			PLUGIN_PROVIDE(PUBKEY, KEY_ML_DSA_65),
				ML_DSA_PUBKEY_DEPS,
...
		PLUGIN_REGISTER(PRIVKEY, ml_dsa_private_key_load, TRUE),
			PLUGIN_PROVIDE(PRIVKEY, KEY_ML_DSA_44),
				ML_DSA_PRIVKEY_DEPS,
...
		PLUGIN_REGISTER(PRIVKEY_GEN, ml_dsa_private_key_gen, FALSE),
			PLUGIN_PROVIDE(PRIVKEY_GEN, KEY_ML_DSA_44),
				ML_DSA_PRIVKEY_GEN_DEPS,
...
		PLUGIN_PROVIDE(PRIVKEY_SIGN, SIGN_ML_DSA_44),
			ML_DSA_PRIVKEY_DEPS,
...
		PLUGIN_PROVIDE(PUBKEY_VERIFY, SIGN_ML_DSA_44),
			ML_DSA_PUBKEY_DEPS
```

tobiasbrunner replied:
I've pushed this as a separate fixup commit to the branch. Let me know what you think.

tobiasbrunner commented on src/libstrongswan/plugins/ml/ml_utils.c:
```diff
@@ -53,3 +53,75 @@ void ml_write_bytes_le(uint8_t *buf, size_t len, uint32_t val)
+/*
+ * Described in header
+ */
+int32_t ml_use_hint(int32_t a, int32_t hint,  int32_t gamma2)
```
```suggestion
int32_t ml_use_hint(int32_t a, int32_t hint, int32_t gamma2)
```

strongX509 replied:
Fixed

tobiasbrunner commented on src/libstrongswan/plugins/ml/ml_utils.c:
```diff
@@ -53,3 +53,75 @@ void ml_write_bytes_le(uint8_t *buf, size_t len, uint32_t val)
+/*
+ * Described in header
+ */
+int32_t ml_make_hint(int32_t a0, int32_t a1,  int32_t gamma2)
```
```suggestion
int32_t ml_make_hint(int32_t a0, int32_t a1, int32_t gamma2)
```

strongX509 replied:
Fixed

tobiasbrunner commented on src/libstrongswan/plugins/ml/ml_utils.h:
```diff
@@ -66,4 +81,41 @@ uint32_t ml_read_bytes_le(uint8_t *buf, size_t len);
+/**
+ * Decompose a into (a1, a0) such that a ≡ a1 * (2 * gamma2) + a0 mod q.
+ *
+ * Algorithm 36 of FIPS 204
```
```suggestion
 * Algorithm 36 of FIPS 204.
```

strongX509 replied:
Fixed

tobiasbrunner commented on src/libstrongswan/plugins/ml/ml_utils.h:
```diff
@@ -66,4 +81,41 @@ uint32_t ml_read_bytes_le(uint8_t *buf, size_t len);
+/**
+ * Return the high bits a1 of a adjusted according to hint h.
+ *
+ * Algorithm 40 of FIPS 204
```
```suggestion
 * Algorithm 40 of FIPS 204.
```

strongX509 replied:
Fixed

tobiasbrunner commented on src/libstrongswan/plugins/ml/ml_utils.h:
```diff
@@ -66,4 +81,41 @@ uint32_t ml_read_bytes_le(uint8_t *buf, size_t len);
+ * @param gamma2	parameter gamma2
+ * @return			hint bit (0 or 1)
+ */
+int32_t ml_make_hint(int32_t a0, int32_t a1,  int32_t gamma2);
```
```suggestion
int32_t ml_make_hint(int32_t a0, int32_t a1, int32_t gamma2);
```

strongX509 replied:
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_params.c:
```diff
@@ -0,0 +1,87 @@
+ * Parameter sets for ML-DSA.
+ */
+static const ml_dsa_params_t ml_dsa_params[] = {
+	{	.type = KEY_ML_DSA_44,
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_params.c:
```diff
@@ -0,0 +1,87 @@
+		.privkey_len = 2560,
+		.sig_len = 2420,
+	},
+	{	.type = KEY_ML_DSA_65,
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+#include "ml_utils.h"
+
+/**
+ * Precalculated Zeta^BitRev_8(i) mod q values for NTT Algorithms 411 and 42
+ * (see Appendix B in FIPS 204).
```
Added comment

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+ */
+void ml_dsa_poly_inv_ntt(ml_dsa_poly_t *a)
+{
+	u_int start, len, j, k = 256;
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+	u_int start, len, j, k = 256;
+	int32_t t, zeta;
+
+	/* scaling factor */
+	const int32_t factor = 41978; /* mont^2/256 */
```
Why should we be holier than the PQ-Crystals-Dilithium project where the code comes from? The don't give any explanations either. From my experience with Bliss and Newhope, there are innumerable formats of the NTT coefficients due to pre-scaling (with or without Montgomery reductions) and/or time or frequency domain shifts in order to minimize the number of operations.

strongX509 replied:
Added comment anyway

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+{
+	while (k--)
+	{
+		b[k] = a[k];
```
Anything with a fixed size can be copied in C. Most prominent example is chunk_t.

tobiasbrunner replied:
> Anything with a fixed size can be copied in C.

Kinda, because you can't just copy plain arrays, even when they have the same fixed size. This apparently only works if they are part of a struct.


> Most prominent example is chunk_t.

Yes, structs can be assigned (we use that e.g. in the `INIT` macro). But those have generally just simple types as members (for `chunk_t` a `size_t` and a `void*`) and initialization could also be different (plain arrays can be initialized via assignment, but not copied). I guess in a struct a fixed size array is just part of the whole structure that's copied (`sizeof()` works correctly as well).

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+/*
+ * Described in header
+ */
+void ml_dsa_poly_add_vec(u_int k, ml_dsa_poly_t *a, ml_dsa_poly_t *b,
+								  ml_dsa_poly_t *res)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+/*
+ * Described in header
+ */
+void ml_dsa_poly_sub_vec(u_int k, ml_dsa_poly_t *a, ml_dsa_poly_t *b,
+								  ml_dsa_poly_t *res)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+/*
+ * Described in header
+ */
+void ml_dsa_poly_mult_mat(u_int k, u_int l, ml_dsa_poly_t *a, ml_dsa_poly_t *b,
+											ml_dsa_poly_t *res)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+/*
+ * Described in header
+ */
+void ml_dsa_poly_power2round_vec(uint k, ml_dsa_poly_t *a,  ml_dsa_poly_t *a0,
+										 ml_dsa_poly_t *a1)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+/*
+ * Described in header
+ */
+void ml_dsa_poly_decompose_vec(uint k, ml_dsa_poly_t *a,  ml_dsa_poly_t *a0,
+									   ml_dsa_poly_t *a1, int32_t gamma2)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+/*
+ * Described in header
+ */
+void ml_dsa_poly_use_hint_vec(uint k, ml_dsa_poly_t *a,  ml_dsa_poly_t *h,
+									  ml_dsa_poly_t *a1, int32_t gamma2)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+
+	while (k--)
+	{
+
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+
+	while (k--)
+	{
+
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+/*
+ * Described in header
+ */
+u_int ml_dsa_poly_make_hint_vec(uint k, ml_dsa_poly_t *a0,  ml_dsa_poly_t *a1,
+									   	ml_dsa_poly_t *h, int32_t gamma2)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+
+	while (k--)
+	{
+
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+
+	for (n = 0; n < ML_DSA_N; n++)
+	{
+		t = a->f[n] >> 31;
+		t = a->f[n] - (t & 2*a->f[n]);
```
Added comment

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.c:
```diff
@@ -0,0 +1,399 @@
+	return TRUE;
+}
+
+bool ml_dsa_poly_check_bound_vec(u_int k, ml_dsa_poly_t *a, int32_t bound)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+/**
+ * Computes the inverse NTT including scaling
+ *
+ * Algorithm 42 in FIPS 204.
+ */
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param b		vector of polynomials b
+ * @param res	vector of resulting polynomials (can be one of the others)
+ */
+void ml_dsa_poly_add_vec(u_int k, ml_dsa_poly_t *a, ml_dsa_poly_t *b,
+								  ml_dsa_poly_t *res);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+								  ml_dsa_poly_t *res);
+
+/**
+ * Subtract polynomials in vector a and b (a[i] +-b[i] mod q for i in 0 to k-1).
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param b		vector of polynomials b
+ * @param res	vector of resulting polynomials (can be one of the others)
+ */
+void ml_dsa_poly_sub_vec(u_int k, ml_dsa_poly_t *a, ml_dsa_poly_t *b,
+								  ml_dsa_poly_t *res);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param b		vector of polynomials b
+ * @param res	result vector of polynomials
+ */
+void ml_dsa_poly_mult_const_vec(u_int k, ml_dsa_poly_t *a, ml_dsa_poly_t *b,
+										 ml_dsa_poly_t *res);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param b		vector of polynomials b
+ * @param res	result polynomial
+ */
+void ml_dsa_poly_mult_vec(u_int k, ml_dsa_poly_t *a, ml_dsa_poly_t *b,
+								   ml_dsa_poly_t *res);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+/**
+ * Dot product of a matrix a with a vector b.
+ *
+ * @param k		number of lines in matrix a
+ * @param l		number of columns in matrix a
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+void ml_dsa_poly_reduce_vec(u_int k, ml_dsa_poly_t *a);
+
+/**
+ * Add q if coefficient is negative
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param a0	vector of polynomials a0
+ * @param a1	vector of polynomials a1
+ */
+void ml_dsa_poly_power2round_vec(uint k, ml_dsa_poly_t *a,  ml_dsa_poly_t *a0,
+										 ml_dsa_poly_t *a1);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+										 ml_dsa_poly_t *a1);
+
+/**
+ * Multiply polynomial by 2^d
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param a1		vector of polynomials a1
+ * @param gamma2	parameter gamma2
+*/
+void ml_dsa_poly_decompose_vec(uint k, ml_dsa_poly_t *a,  ml_dsa_poly_t *a0,
+									   ml_dsa_poly_t *a1, int32_t gamma2);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param a1		vector of polynomials a1
+ * @param gamma2	parameter gamma2
+*/
+void ml_dsa_poly_use_hint_vec(uint k, ml_dsa_poly_t *a,  ml_dsa_poly_t *h,
+									  ml_dsa_poly_t *a1, int32_t gamma2);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_poly.h:
```diff
@@ -0,0 +1,238 @@
+ * @param gamma2	parameter gamma2
+ * @return			total numer of hint bits
+ */
+u_int ml_dsa_poly_make_hint_vec(uint k, ml_dsa_poly_t *a0,  ml_dsa_poly_t *a1,
+									   	ml_dsa_poly_t *h, int32_t gamma2);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+/* from ml_dsa_public_key.c */
+chunk_t ml_dsa_get_context(void);
+
+bool ml_dsa_expand_a(const ml_dsa_params_t *params, xof_t *G, chunk_t rho,
+						ml_dsa_poly_t *a);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	memcpy(tr.ptr, sk.ptr, tr.len);
+	sk = chunk_skip(sk, tr.len);
+
+	/* unpack the vectors s1, s2 and t0 from the private key blob*/
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+{
+	const u_int k = this->params->k;
+	const u_int l = this->params->l;
+	const u_int d = this->params->d;
```
This constants are usually used in for loops where the loop variable is defined as an u_int. This possibly might avoid some casting (Never know about the capability of modern compilers).

tobiasbrunner replied:
Sorry, I probably put this comment in the wrong place. It's not about `d` being a `u_int` here, it's about `d` in `ml_dsa_params_t` being defined as a `size_t`, while all the others are `uint8_t`. This causes a warning here in SonarQube due to the implicit downcast from `size_t` to `u_int` (the upcast from `uint8_t` to `u_int` is fine). I've pushed a fix to the branch.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+			{
+				goto end;
+			}
+			t0[i].f[n] = (1 << (ML_DSA_D-1)) - value;
```
Added cast for the sake of consistency.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+		rho.ptr[ML_DSA_RHO_PP_LEN+1] = nonce++ >> 8;
+
+		if (!this->H->set_seed(this->H, rho) ||
+			!this->H->get_bytes(this->H, 32*(1 + gamma1_exp), v.ptr))
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+}
+
+/**
+ * encodes a signature into a byte string (sigEncode)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	{
+		for (n = 0; n < ML_DSA_N; n++)
+		{
+			if (!bitpacker->write_bits(bitpacker, (1 << gamma1_exp) - z[j].f[n],
+												   1 +  gamma1_exp))
```
Cosmetics

tobiasbrunner replied:
I guess, but it really doesn't make that much sense to align it this way. These are separate calculations and function arguments (the only connection is that they both use the same constant). This way it gives the initial impression that the second line was indented like that because it's a continuation of the calculation on the previous line (especially because it was aligned to the opening bracket there).

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	chunk_t signature, sig;
+	u_int i, j, n, index = 0;
+
+	signature= chunk_alloc(this->params->sig_len);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+
+	/* encode vector of polynomials z in packed format */
+	bitpacker = ml_bitpacker_create(sig);
+	for (j = 0; j < l; j++)
```
Looping up to l consistently uses the j loop variable wheras looping up to k uses the i loop variable.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	if (!decode_secret_key(this, rho, K, tr, s1, s2, t0) ||
+		!ml_dsa_expand_a(this->params, this->G, rho, a))
+	{
+		return FALSE;
```
Well if ml_dsa_expand_a() fails then everything is rotten anyway and there is no special reason to protect the private key.

But you are right. In principle the ml_poly_t arrays s1, s2 and t0 comprising the private key should be memwiped before exiting.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	if (!this->H->set_seed(this->H, seed) ||
+		!this->H->get_bytes(this->H, ML_DSA_RHO_PP_LEN, rho_pp.ptr))
+	{
+		chunk_free(&seed);
```
Hmm, this is not much of a secret since it just randomizes the signature. The static secret K might be counted as part of the private key, though. Thus if in doubt wipe the seed.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+		ml_dsa_poly_inv_ntt_vec(k, w1);
+		ml_dsa_poly_cond_add_q_vec(k, w1);
+
+		/* deoompose elements of vector w into high and low bits */
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	ml_dsa_poly_t a[k*l], s1[l], s2[k], t0[k];
+	ml_dsa_poly_t y[l], z[l], w0[k], w1[k], h[k], c;
+	chunk_t rho, K , tr, ctx, seed, mu, rnd, rho_pp, w1_enc, c_tilde;
+	int32_t bound1 = (1 << this->params->gamma1_exp) - this->params->beta;
```
Changed to constant and joined with others

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	ml_dsa_poly_t y[l], z[l], w0[k], w1[k], h[k], c;
+	chunk_t rho, K , tr, ctx, seed, mu, rnd, rho_pp, w1_enc, c_tilde;
+	int32_t bound1 = (1 << this->params->gamma1_exp) - this->params->beta;
+	int32_t bound2 = gamma2 - this->params->beta;
```
Changed to constant and joined with others

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+		if (!this->H->set_seed(this->H, seed) ||
+			!this->H->get_bytes(this->H, this->params->lambda/4, c_tilde.ptr))
+		{
+			chunk_free(&seed);
```
If in doubt yes

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+		DESTROY_IF(this->G);
+		DESTROY_IF(this->H);
+		chunk_clear(&this->keyseed);
+		chunk_free(&this->pubkey);
+		chunk_clear(&this->privkey);
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	{
+		return FALSE;
+	}
+	memwipe(seedbuf, sizeof(seedbuf));
```
memwipe on the sensitivy ml_dsa_poly_t a[], s1[], s2[], and t0[] as well as chunk_t K forming part of each signature.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	 */
+	uint8_t seedbuf[4*ML_DSA_SEED_LEN];
+	chunk_t seed =     { seedbuf,                       ML_DSA_SEED_LEN+2 };
+	chunk_t rho  =     { seedbuf,                       ML_DSA_SEED_LEN};
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+static bool expand_s(private_private_key_t *this, chunk_t rhoprime,
+					 ml_dsa_poly_t *s1, ml_dsa_poly_t *s2)
+{
+	chunk_t seed;;
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	chunk_t seed;;
+	const u_int k = this->params->k;
+	const u_int l = this->params->l;
+	u_int i,j;
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	const u_int l = this->params->l;
+	u_int i,j;
+
+	seed = chunk_alloca(2*ML_DSA_SEED_LEN + 2);
```
wiped

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+		{
+			if (++n == ML_DSA_N)
+			{
+				return TRUE;
```
fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+		{
+			if (!bitpacker->write_bits(bitpacker, t1[i].f[n], ML_DSA_T1_BITS))
+			{
+				goto end;
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+
+/**
+ * Samples vectors s1 and s2, each with polynomial coordinates whose coefficients
+ * are in the interval [-eta, eta]
```
FIxed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+}
+
+/**
+ * Encode the secret key (skEncode)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	memcpy(sk.ptr, K.ptr, K.len);
+	sk = chunk_skip(sk, K.len);
+
+	/**
+	 * compute tr as a SHAKE-256 digest over the public key blob
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+
+	/* pack the vectors s1, s2 and t0 into the private key blob*/
+	bitpacker = ml_bitpacker_create(sk);
+	for (j = 0; j < l; j++)
```
See above

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	{
+		for (n = 0; n < ML_DSA_N; n++)
+		{
+			if (!bitpacker->write_bits(bitpacker, (1 << (ML_DSA_D-1)) - t0[i].f[n],
+														 ML_DSA_D))
```
I prefer this alignment.

tobiasbrunner replied:
Hm, as mentioned with the other one, this looks confusing as it implicates a continuation of the calculation on the previous line. There is really no point in aligning unrelated function arguments like that.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+		}
+		break;
+	}
+
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	this = create_instance(type);
+	if (!this)
+	{
+		return NULL;
```
Why should we clear the seed when the key generation fails?

tobiasbrunner replied:
I guess it's `RNG_TRUE` so there should be no danger in leaking RNG state, but it's still a memory leak. I've pushed a fix to the branch.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.h:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Load an ML-DSA private key using.
+ *
+ * Accepts a BUILD_BLOB_ASN1_DER argument.
```
Fixed, but with local PKCS#8 parsing, KEY_ANY will also be supported.

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+/*
+  * Copyright (C) 2024 Andreas Steffen, strongSec GmbH
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+#include "ml_dsa_params.h"
+#include "ml_dsa_poly.h"
+#include "ml_bitpacker.h"
+
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+}
+
+/**
+ * decode the public key (pkDncode)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+}
+
+/**
+ * decodes a signature (sigDecode)
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+
+	/* decode vector of polynomials z from packed format */
+	bitpacker = ml_bitpacker_create_from_data(sig);
+	for (j = 0; j < l; j++)
```
j is always used with l

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+			{
+				goto end;
+			}
+			s1[j].f[n] = eta - (int32_t)value;
```
Added range checks

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+			{
+				goto end;
+			}
+			s2[i].f[n] = eta - (int32_t)value;
```
Added range checks

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+	bitpacker->destroy(bitpacker);
+	sig = chunk_skip(sig, 32 * (1 + gamma1_exp) * l);
+
+	/* encode vector of polynomials with binary coefficients h */
```
Mentioned algorithm 20

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+	bitpacker->destroy(bitpacker);
+	sig = chunk_skip(sig, 32 * (1 + gamma1_exp) * l);
+
+	/* decode vector of polynomials with binary coefficients h */
```
Mentioned algorithm 21

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+			return FALSE;
+		}
+
+		/**
+		 *  Algorithm 14 in FIPS 204 (CoeffFromThreeBytes).
+		 */
```
Ok

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+}
+
+/**
+ * Samples a k x l matrix A
```
Fixed

strongX509 commented on src/libstrongswan/plugins/ml/ml_dsa_public_key.c:
```diff
@@ -0,0 +1,713 @@
+		len = strlen(context_str);
+
+		/* context in hex format? */
+		if (len > 2 && context_str[0] == '0' && context_str[1] == 'x')
```
Fixed

strongX509 committed:
testing: Add rw-ml-kem-dsa scenarios

strongX509 commented:
Hmm, der Fix ist irgendwie verlorengegangen. Habe in nun gepusht.

On 06.02.25 10:17, Tobias Brunner wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/pki/commands/gen.c <https://github.com/strongswan/strongswan/ 
> pull/2626#discussion_r1944373671>:
> 
>> +			case KEY_ML_DSA_44:
> +				size = 10496;
> +				break;
> +			case KEY_ML_DSA_65:
> +				size = 15616;
> +				break;
> +			case KEY_ML_DSA_87:
> +				size = 20736;
> +				break;
> 
> It's still there :) I'll push this and some other fixups to the branch 
> for your consideration.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> strongswan/strongswan/pull/2626#discussion_r1944373671>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> ABFV774TKSD2DNOOY3GLAYT2OMSELAVCNFSM6AAAAABVQJRT4KVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOJYGA4TCNRXHE>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
> 

-- 
======================================================================
Andreas Steffen                         ***@***.***
strongSwan - the Open Source VPN Solution!          www.strongswan.org
strongSec GmbH, 8952 Schlieren (Switzerland)
======================================================================

tobiasbrunner commented on src/libstrongswan/plugins/ml/ml_dsa_private_key.c:
```diff
@@ -0,0 +1,995 @@
+
+	*signature = encode_signature(this, c_tilde, z, h);
+
+	return signature->len > 0;
```
I think e.g. `sizeof(a)` should work fine here as well, no need to calculate the size manually for those arrays (same for the `memwipe()` calls in other functions). See my suggestion in the fixup commit.

tobiasbrunner commented:
Just for reference, there is a [00 version of a draft](https://datatracker.ietf.org/doc/html/draft-sfluhrer-ipsecme-ikev2-mldsa-00) that describes how to use ML-DSA with IKEv2. It currently differs from this implementation in several aspects that are less than ideal and hopefully will change in the future:

 1. It uses an IKEv2-specific context string, which is problematic in two ways. First, context strings are not supported by all implementations yet. Second, this string would have to be passed as param and not a global setting, because the latter would conflict with certificate chain validation that likely won't use that same context string (the implementations don't know if they are signing/validating for IKEv2 authentication or a certificate chain so two global settings aren't really a solution either).
 2. While the draft also intends to use the pure version of ML-DSA (as compared to the prehashed version) it defines a "fake" prehash for the auth octets before signing, instead of just using RFC 8420's identity hash (it does discuss the latter, though. The draft is also suggesting that the selected hash algorithm could be retrieved from the algorithm identifier, but it's unclear how that would be encoded.

strongX509 committed:
openssl: Add ML-DSA support

strongX509 committed:
leak_detective: Add OSSL_PROVIDER_add_conf_parameter to whitelist

SebastienRaveau commented:
Hi! Not sure where to report that [the strongSwan wiki is down](https://downforeveryoneorjustme.com/wiki.strongswan.org?proto=https) so may I seize the opportunity to ask if there is anything I can do to help with this branch/PR?

I have been upgrading as much of my configs/code as possible to PQC KEM & DSA, perhaps some of the [debugging experience](https://github.com/rustls/rustls/issues/2577) could be reused :slightly_smiling_face:

strongX509 committed:
unit-tests: Test all ML-DSA private key formats

strongX509 committed:
ml: Support all ML-DSA private key formats

tobiasbrunner committed:
openssl: Add generic loader for public keys to avoid pkcs1 dependency

This can load public keys of unknown type via KEY_ANY without requiring
the pkcs1 plugin.

tobiasbrunner committed:
signature-params: Only encode length in context string of PQC params

While the length prefix is also used for PQC or hybrid algorithms other
than ML-DSA, the first byte that was previously encoded is used to
distinguish ML-DSA from HashML-DSA (Pre-Hash ML-DSA), where it would be
set to 1.  NIST doesn't recommend that mode and we currently don't
support it.  If we ever do, that flag could be made part of the params.

tobiasbrunner committed:
ml: Support encoding ML-DSA keys in expanded format if we don't have the seed

The previous code failed to produce a valid encoding so e.g. passing
an encrypted key loaded from the expanded format couldn't get loaded
via VICI.

The tests are expanded to check that we get the expected encoding (i.e.
seed-only except if we loaded a key in expanded format).

tobiasbrunner committed:
github: Use OpenSSL 3.6.3 and 4.0.1 for tests

tobiasbrunner committed:
wolfssl: Adapt to removed ML-KEM header

The mlkem.h header that mainly defined aliases for the old wc_Kyber* API
has been removed and its contents moved to the wc_mlkem.h header.

tobiasbrunner committed:
Use wolfSSL 5.9.2 for tests
